Enable pyupgrade as a pants backend

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
Fixes https://github.com/grapl-security/issue-tracker/issues/971
<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
This adds a new tool called [pyupgrade](https://github.com/asottile/pyupgrade), which will rewrite python code to newer python syntax. This has been added as part of the pants suite, and will make the upgrade to python 3.10 MUCH easier
<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
```
./pants lint ::
```
<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Enable pyupgrade as a pants backend

[pyupgrade](https://github.com/asottile/pyupgrade) is a tool that
rewrites python code to adhere to newer versions of python. This makes
upgrading python major versions much easier and reduces the toil in
doing so.

colin-grapl commented on pants.toml:
```diff
@@ -207,6 +208,9 @@ known_versions = [
 ]
 
+[pyupgrade]
+args = ["--py38-plus"]
```
I think we may be on 3.7 in some places, like CI. We should update those first.

christophermaier replied:
Things in CI are set up to honor `.python-version`, so we should be good.

@colin-grapl What specific instances of 3.7 are you seeing and concerned about?

christophermaier reviewed (approved):
Nice addition... worth the price of admission for the `Optional[X]` -> `X | None` update alone :joy:

colin-grapl commented on etc/ci_scripts/dump_artifacts/nomad_artifacts.py:
```diff
@@ -69,7 +69,7 @@ def _dump_hashicorp_agent_logs(artifacts_dir: Path) -> None:
     name: str
 
-    def __init__(self, input: Dict[str, Any]) -> None:
+    def __init__(self, input: dict[str, Any]) -> None:
```
Same here, I've always assumed that `List` and `Dict` were preferred to `list` and `dict` but I could absolutely be wrong

twunderlich-grapl commented on etc/ci_scripts/dump_artifacts/nomad_artifacts.py:
```diff
@@ -18,14 +18,14 @@
 
 OutOrErr = Literal["stdout", "stderr"]
-OUTPUT_TYPES: List[OutOrErr] = ["stdout", "stderr"]
+OUTPUT_TYPES: list[OutOrErr] = ["stdout", "stderr"]
```
So this is apparently preferred per [PEP 585](https://peps.python.org/pep-0585/#abstract). Which honestly after skimming makes sense

Commit:
Remove unused imports

Its probably close to time to set up autoflake to make this faster

Commit:
Update remaining python 3.7 references to 3.8

Commit:
Add back the Dict and List that was manually changed

Commit:
Upgrade pulumi container to python 3.10

We're running into an error triggered only by the integration tests
because the GenericAlias type only added Union support as of 3.10 per
https://bugs.python.org/issue42233. Since we're planning on upgrading
to 3.10 shortly anyway, this should be fine (ie the upgrade should be
within the next week)

christophermaier commented on Dockerfile.pulumi:
```diff
@@ -1,4 +1,6 @@
-FROM python:3.8-slim-bullseye as grapl-local-pulumi
+FROM python:3.10-slim-bullseye as grapl-local-pulumi
```
Once we've got everything on 3.10 --- that is, when our `.python-version` file contains a 3.10.x version --- then we could rig something up so that we pass that version in as a build argument and incorporate that into the base image identifier. That would be one less thing we'd have to keep in sync. (That goes for the `src/python/Dockerfile`, as well.)

(That's not a change for this PR, of course... just mentioning the idea for the future.)

christophermaier replied:
Prior art for Rust: https://github.com/grapl-security/grapl/pull/1788/commits/bdefc4d1ffeb5fade35ae203d04f0aa113feeec2

christophermaier replied:
Actually, let's just make an issue, shall we? 😅 https://github.com/grapl-security/issue-tracker/issues/973

Commit:
Merge branch 'main' into twunderlich/pants-pyupgrade

Commit:
Remove the pantsbuild dependencies
